Crash in Fotoapparat.start(): IOException: setPreviewTexture failed

#### What are you trying to achieve or the steps to reproduce?
This crash was observed on a Samsung Galaxy Grand Prime running Android 5.1.1. It occurred when the camera activity was destroyed immediately after being resumed. Here is the scenario:
1. CameraActivity takes a picture and launches an ImageReviewActivity.
2. If the user accepts the image, ImageReviewActivity closes and returns success to CameraActivity.
3. CameraActivity resumes, save the image to disk, and exits immediately.

During step 3 above, CameraActivity's onStart() (calls camera.start()),  onStop() (call camera.stop()), and onDestroy() are all called in less than 1 second.

#### What was the result you received?
Fatal crash due to an uncaught IOException: setPreviewTexture failed in Fotoapparat's startRoutine.

**Logs:**
> 01-11 14:31:34.480 9246-9246 D/Fotoapparat: Fotoapparat: start
> 01-11 14:31:34.480 9246-11084 D/Fotoapparat: Device: selectCamera
> 01-11 14:31:34.480 9246-9246 D/Fotoapparat: Fotoapparat: getCurrentParameters
> 01-11 14:31:34.480 9246-11084 D/Fotoapparat: CameraDevice: open
> 01-11 14:31:34.480 9246-9246 D/Fotoapparat: Fotoapparat: getCapabilities
> 01-11 14:31:34.490 9246-11679 D/Fotoapparat: CameraDevice: getParameters$suspendImpl
> 01-11 14:31:34.490 9246-11679 D/Fotoapparat: CameraDevice: getCapabilities$suspendImpl
> 01-11 14:31:34.630 9246-11084 D/Fotoapparat: CameraDevice: getCapabilities$suspendImpl
> 01-11 14:31:34.630 9246-11084 D/Fotoapparat: CameraDevice: updateParameters$suspendImpl
> 01-11 14:31:34.630 9246-11084 D/Fotoapparat: New camera parameters are: […]
> 01-11 14:31:34.640 9246-11084 D/Fotoapparat: CameraDevice: updateFrameProcessor
> 01-11 14:31:34.640 9246-11084 D/Fotoapparat: CameraDevice: setDisplayOrientation
> 01-11 14:31:34.640 9246-11084 D/Fotoapparat: Image Rotation is: 270. Display rotation is: 90
> 01-11 14:31:34.650 9246-11084 D/Fotoapparat: CameraDevice: getPreviewResolution
> 01-11 14:31:34.650 9246-11084 D/Fotoapparat: Preview resolution is: Resolution(width=720, height=1280)
> 01-11 14:31:34.650 9246-11084 D/Fotoapparat: CameraDevice: setDisplaySurface
> 01-11 14:31:34.650 9246-11084 E/BufferQueueProducer: [unnamed-9246-0] connect(P): BufferQueue has been abandoned
> 01-11 14:31:34.650 9246-11084 E/AndroidRuntime: FATAL EXCEPTION: pool-4-thread-1
>                                                                          Process: [redacted], PID: 9246
>                                                                          java.lang.Error: java.io.IOException: setPreviewTexture failed
>                                                                              at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1118)
>                                                                              at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:587)
>                                                                              at java.lang.Thread.run(Thread.java:818)
>                                                                           Caused by: java.io.IOException: setPreviewTexture failed
>                                                                              at android.hardware.Camera.setPreviewTexture(Native Method)
>                                                                              at io.fotoapparat.hardware.CameraDeviceKt.setDisplaySurface(CameraDevice.kt:402)
>                                                                              at io.fotoapparat.hardware.CameraDeviceKt.access$setDisplaySurface(CameraDevice.kt:1)
>                                                                              at io.fotoapparat.hardware.CameraDevice.setDisplaySurface(CameraDevice.kt:251)
>                                                                              at io.fotoapparat.routine.camera.StartRoutineKt.start(StartRoutine.kt:67)
>                                                                              at io.fotoapparat.routine.camera.StartRoutineKt.bootStart(StartRoutine.kt:22)
>                                                                              at io.fotoapparat.Fotoapparat$start$1.invoke(Fotoapparat.kt:89)
>                                                                              at io.fotoapparat.Fotoapparat$start$1.invoke(Fotoapparat.kt:41)
>                                                                              at io.fotoapparat.hardware.ExecutorKt$sam$Runnable$ffbfbfd0.run(Executor.kt)
>                                                                              at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1112)
>                                                                              at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:587) 
>                                                                              at java.lang.Thread.run(Thread.java:818) 

#### What did you expect?
If there is an exception inside Fotoapparat's async routines, it should be caught and, if it's fatal, be sent to the cameraErrorCallback. 
I could work around this by optimizing the code not to call camera.start() when it's about to be destroyed but I still wanted to report this uncaught exception since it may affect others.


#### Context:

* *FA version*: 2.1.2
* *Devices/APIs affected*: Samsung Galaxy Grand Prime running Android 5.1.1

I have the same bug here : 
```
E/AndroidRuntime: FATAL EXCEPTION: pool-2-thread-1
                                                                                     Process: ma.ensamcasa.objectdetectionandroid, PID: 16702
                                                                                     java.lang.Error: java.io.IOException: setPreviewTexture failed
                                                                                         at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1119)
                                                                                         at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:588)
                                                                                         at java.lang.Thread.run(Thread.java:818)
                                                                                      Caused by: java.io.IOException: setPreviewTexture failed
                                                                                         at android.hardware.Camera.setPreviewTexture(Native Method)
                                                                                         at io.fotoapparat.hardware.CameraDeviceKt.setDisplaySurface(CameraDevice.kt:402)
                                                                                         at io.fotoapparat.hardware.CameraDeviceKt.access$setDisplaySurface(CameraDevice.kt:1)
                                                                                         at io.fotoapparat.hardware.CameraDevice.setDisplaySurface(CameraDevice.kt:251)
                                                                                         at io.fotoapparat.routine.camera.StartRoutineKt.start(StartRoutine.kt:67)
                                                                                         at io.fotoapparat.routine.camera.StartRoutineKt.bootStart(StartRoutine.kt:22)
                                                                                         at io.fotoapparat.Fotoapparat$start$1.invoke(Fotoapparat.kt:89)
                                                                                         at io.fotoapparat.Fotoapparat$start$1.invoke(Fotoapparat.kt:41)
                                                                                         at io.fotoapparat.hardware.ExecutorKt$sam$Runnable$ffbfbfd0.run(Executor.kt)
                                                                                         at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1113)
                                                                                         at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:588) 
                                                                                         at java.lang.Thread.run(Thread.java:818) 
```
this problem happens when  I hit back button in any second activity to return to cameraactivity.
Problem happened in Emulator android api level 23

I have the same bug here : 
```
E/AndroidRuntime: FATAL EXCEPTION: pool-2-thread-1
                                                                                     Process: .... PID: 16702
                                                                                     java.lang.Error: java.io.IOException: setPreviewTexture failed
                                                                                         at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1119)
                                                                                         at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:588)
                                                                                         at java.lang.Thread.run(Thread.java:818)
                                                                                      Caused by: java.io.IOException: setPreviewTexture failed
                                                                                         at android.hardware.Camera.setPreviewTexture(Native Method)
                                                                                         at io.fotoapparat.hardware.CameraDeviceKt.setDisplaySurface(CameraDevice.kt:402)
                                                                                         at io.fotoapparat.hardware.CameraDeviceKt.access$setDisplaySurface(CameraDevice.kt:1)
                                                                                         at io.fotoapparat.hardware.CameraDevice.setDisplaySurface(CameraDevice.kt:251)
                                                                                         at io.fotoapparat.routine.camera.StartRoutineKt.start(StartRoutine.kt:67)
                                                                                         at io.fotoapparat.routine.camera.StartRoutineKt.bootStart(StartRoutine.kt:22)
                                                                                         at io.fotoapparat.Fotoapparat$start$1.invoke(Fotoapparat.kt:89)
                                                                                         at io.fotoapparat.Fotoapparat$start$1.invoke(Fotoapparat.kt:41)
                                                                                         at io.fotoapparat.hardware.ExecutorKt$sam$Runnable$ffbfbfd0.run(Executor.kt)
                                                                                         at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1113)
                                                                                         at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:588) 
                                                                                         at java.lang.Thread.run(Thread.java:818) 
```
this problem happens when  I hit back button in any second activity to return to cameraactivity.
Problem happened in Emulator android api level 23

I can also reproduce this error with the following code
```
public class MainActivity extends AppCompatActivity
{
	private Fotoapparat fotoapparat;

	@Override
	protected void onCreate(Bundle savedInstanceState)
	{
		super.onCreate(savedInstanceState);
		setContentView(R.layout.activity_main);
		CameraView cameraView = findViewById(R.id.camera_view);
		fotoapparat = Fotoapparat
			.with(this)
			.into(cameraView)
			.lensPosition(back())
			.focusMode(firstAvailable(
				continuousFocusPicture(),
				autoFocus(),
				fixed()
			))
			.flash(off())
			.build();
	}

	@Override
	protected void onResume()
	{
		super.onResume();
		fotoapparat.start();
	}

	@Override
	protected void onPause()
	{
		fotoapparat.stop();
		super.onPause();
	}
}
```
When pausing/resuming the app really fast a couple of times on a Galaxy S6 running Android 6.0.1 I get the following crash:

```
 Process: com.navvis.fotoapparattest, PID: 24056
  java.lang.Error: java.io.IOException: setPreviewTexture failed
      at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1119)
      at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:588)
      at java.lang.Thread.run(Thread.java:818)
   Caused by: java.io.IOException: setPreviewTexture failed
      at android.hardware.Camera.setPreviewTexture(Native Method)
      at io.fotoapparat.hardware.CameraDeviceKt.setDisplaySurface(CameraDevice.kt:402)
      at io.fotoapparat.hardware.CameraDeviceKt.access$setDisplaySurface(CameraDevice.kt:1)
      at io.fotoapparat.hardware.CameraDevice.setDisplaySurface(CameraDevice.kt:251)
      at io.fotoapparat.routine.camera.StartRoutineKt.start(StartRoutine.kt:67)
      at io.fotoapparat.routine.camera.StartRoutineKt.bootStart(StartRoutine.kt:22)
      at io.fotoapparat.Fotoapparat$start$1.invoke(Fotoapparat.kt:89)
      at io.fotoapparat.Fotoapparat$start$1.invoke(Fotoapparat.kt:41)
      at io.fotoapparat.hardware.ExecutorKt$sam$Runnable$ffbfbfd0.run(Executor.kt)
      at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1113)
      at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:588) 
      at java.lang.Thread.run(Thread.java:818) 
```
.

--- a/fotoapparat/src/main/java/io/fotoapparat/hardware/CameraDevice.kt
+++ b/fotoapparat/src/main/java/io/fotoapparat/hardware/CameraDevice.kt
@@ -258,6 +258,7 @@ internal open class CameraDevice(
     /**
      * Sets the desired surface on which the camera's preview will be displayed.
      */
+    @Throws(IOException::class)
     open fun setDisplaySurface(preview: Preview) {
         logger.recordMethod()
 

--- a/fotoapparat/src/main/java/io/fotoapparat/hardware/Device.kt
+++ b/fotoapparat/src/main/java/io/fotoapparat/hardware/Device.kt
@@ -25,7 +25,7 @@ import kotlinx.coroutines.experimental.CompletableDeferred
  * Phone.
  */
 internal open class Device(
-        private val logger: Logger,
+        internal open val logger: Logger,
         private val display: Display,
         internal open val scaleType: ScaleType,
         internal open val cameraRenderer: CameraRenderer,

--- a/fotoapparat/src/main/java/io/fotoapparat/routine/camera/StartRoutine.kt
+++ b/fotoapparat/src/main/java/io/fotoapparat/routine/camera/StartRoutine.kt
@@ -9,6 +9,7 @@ import io.fotoapparat.hardware.orientation.OrientationSensor
 import io.fotoapparat.hardware.orientation.OrientationState
 import io.fotoapparat.routine.focus.focusOnPoint
 import io.fotoapparat.routine.orientation.startOrientationMonitoring
+import java.io.IOException
 
 /**
  * Starts the camera from idle.
@@ -69,11 +70,15 @@ internal fun Device.start() {
         })
     }
 
-    cameraDevice.run {
-        setDisplaySurface(
-                preview = cameraRenderer.getPreview()
-        )
+    with(cameraDevice) {
+        try {
+            setDisplaySurface(
+                    preview = cameraRenderer.getPreview()
+            )
 
-        startPreview()
+            startPreview()
+        } catch (e: IOException) {
+            logger.log("Can't start preview because of the exception: $e")
+        }
     }
 }

--- a/fotoapparat/src/test/java/io/fotoapparat/routine/camera/StartRoutineTest.kt
+++ b/fotoapparat/src/test/java/io/fotoapparat/routine/camera/StartRoutineTest.kt
@@ -8,6 +8,7 @@ import io.fotoapparat.hardware.orientation.Orientation.Horizontal.Landscape
 import io.fotoapparat.hardware.orientation.Orientation.Vertical.Portrait
 import io.fotoapparat.hardware.orientation.OrientationSensor
 import io.fotoapparat.hardware.orientation.OrientationState
+import io.fotoapparat.log.Logger
 import io.fotoapparat.parameter.ScaleType
 import io.fotoapparat.test.testResolution
 import io.fotoapparat.test.willReturn
@@ -15,14 +16,13 @@ import io.fotoapparat.test.willThrow
 import io.fotoapparat.view.CameraRenderer
 import io.fotoapparat.view.Preview
 import io.fotoapparat.view.toPreview
-import kotlinx.coroutines.experimental.runBlocking
 import org.junit.Before
 import org.junit.Test
 import org.junit.runner.RunWith
 import org.mockito.Mock
-import org.mockito.Mockito.inOrder
-import org.mockito.Mockito.verify
+import org.mockito.Mockito.*
 import org.mockito.junit.MockitoJUnitRunner
+import java.io.IOException
 import java.util.concurrent.atomic.AtomicBoolean
 import kotlin.test.assertFalse
 import kotlin.test.assertTrue
@@ -39,6 +39,8 @@ internal class StartRoutineTest {
     lateinit var orientationSensor: OrientationSensor
     @Mock
     lateinit var cameraDevice: CameraDevice
+    @Mock
+    lateinit var mockLogger: Logger
     @Mock
     lateinit var device: Device
 
@@ -86,6 +88,27 @@ internal class StartRoutineTest {
         }
     }
 
+    @Test
+    fun `Cannot set display surface while starting`() {
+        // Given
+        device.apply {
+            getSelectedCamera() willReturn cameraDevice
+            getScreenOrientation() willReturn Landscape
+            cameraRenderer willReturn cameraViewRenderer
+            scaleType willReturn ScaleType.CenterCrop
+            logger willReturn mockLogger
+        }
+        cameraDevice.getPreviewResolution() willReturn testResolution
+        cameraDevice.setDisplaySurface(preview) willThrow IOException()
+        cameraViewRenderer.getPreview() willReturn preview
+
+        // When
+        device.start()
+
+        // Then
+        verify(cameraDevice, never()).startPreview()
+    }
+
     @Test(expected = IllegalStateException::class)
     fun `Boot start, but camera has already started`() {
         // Given
@@ -118,9 +141,10 @@ internal class StartRoutineTest {
     }
 
     @Test
-    fun `Boot start`() = runBlocking {
+    fun `Boot start`() {
         // Given
-        val hasErrors = AtomicBoolean(false)
+        var hasErrors = false
+
         device.apply {
             hasSelectedCamera() willReturn false
             getSelectedCamera() willReturn cameraDevice
@@ -134,11 +158,11 @@ internal class StartRoutineTest {
         // When
         device.bootStart(
                 orientationSensor,
-                { hasErrors.set(true) }
+                { hasErrors = true }
         )
 
         // Then
-        assertFalse(hasErrors.get())
+        assertFalse(hasErrors)
 
         verify(device).start()
     }

